### PR TITLE
Adds Marching Cubes algorithm to PhysicsNeMo-Mesh

### PR DESCRIPTION
**Pull request by peterdsharpe**
# PhysicsNeMo Pull Request

## Description

Allows mesh generation from SDF files using marching cubes. Part of softening the deprecation blow of #1487.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [x] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/mesh/generate/marching_cubes.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        mc.indices.numpy().reshape(-1, 3), dtype=torch.int64
+    )  # (N_f, 3)
+
+    ### Map from grid-index space to physical coordinates
+    if coords is not None:
+        for dim, c in enumerate(coords):
+            n = c.shape[0]
+            origin = c[0].item()
+            spacing = (c[-1].item() - origin) / (n - 1) if n > 1 else 1.0
+            points[:, dim] = origin + points[:, dim] * spacing
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/mesh/generate/marching_cubes.py]: Fixed in 9cfcde69a32f3ec247d8f663e1eda5ea5be2f690

File: physicsnemo/mesh/generate/marching_cubes.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    mc.surface(field=field_wp, threshold=threshold)
+
+    points = torch.as_tensor(mc.verts.numpy(), dtype=torch.float32)  # (N_v, 3)
+    cells = torch.as_tensor(
+        mc.indices.numpy().reshape(-1, 3), dtype=torch.int64
+    )  # (N_f, 3)
+
+    ### Map from grid-index space to physical coordinates via piecewise linear interp
+    if coords is not None:
+        for dim, c in enumerate(coords):
```
Comment (coreyjadams) [inline physicsnemo/mesh/generate/marching_cubes.py:135]: What is coords here?  Looping over it caught my eye.

**Comment by coreyjadams** [comment]
I don't want to be a nay-sayer, but, do we need this for anything?  If so - and I'm not calling it out as _wrong_ yet, just that I haven't thought about it here - using warp in physicsnemo mesh is a bit of a left-turn, from a design standpoint, that I'm not sure we should make.  It'd be nice to keep warp in the `nn.functional` space if possible, just to keep things clean and isolated.

If there is a compelling reason to build `marching cubes` here in physicsnemo mesh with warp, let's do it, but if it's just to avoid rug pulling on the utils.mesh package I think maybe we can rug pull that one?

File: physicsnemo/mesh/generate/marching_cubes.py (hunk 2 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/generate/marching_cubes.py:135]: Coords is a tuple of length (n_spatial_dims), so 3 in 3D. This is ok

**Comment by peterdsharpe** [comment]
(Summarizing offline discussion)

Added to support geometry encoder work, as well as to soften the blow of `physicsnemo.utils.mesh` deprecation.
